mirgecom/profiling.py

```diff
@@ -0,0 +1,201 @@
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+"""
+
+from meshmode.array_context import PyOpenCLArrayContext
+import pyopencl as cl
+from pytools.py_codegen import PythonFunctionGenerator
+import loopy as lp
+import numpy as np
+
+from loopy.target.execution import ExecutionWrapperGeneratorBase
+
+
+class ProfileData:
+    time = 0
+    flops = 0
+    mem_access = 0
+
+    def __init__(self, time=0, flops=0, mem_access=0):
+        self.time = time
+        self.flops = flops
+        self.mem_access = mem_access
+
+    def __repr__(self):
+        return "(time={0}, flops={1}, mem_access={2})".format(self.time, self.flops, self.mem_access)
+
+    def __str__(self):
+        return self.__repr__()
+
+
+class TimingEvent:
+    def __init__(self, event, program, kwargs):
+        self.event = event
+        self.program = program
+        self.kwargs = kwargs
+
+
+class PyOpenCLProfilingArrayContext(PyOpenCLArrayContext):
+
+    def __init__(self, queue, allocator=None):
+        super().__init__(queue, allocator)
+
+        if not queue.properties & cl.command_queue_properties.PROFILING_ENABLE:
+            from warnings import warn
+            warn("Profiling was not enabled in the command queue. Timing data will not be collected.")
+            self.profiling_enabled = False
+        else:
+            self.profiling_enabled = True
+
+        self.events = []
+        self.profiling_data = {}
+        self.invoker_codes = {}
+
+    def finish_profile_events(self):
+
+        if not self.profiling_enabled:
+            return
+
+        if self.events:
+            cl.wait_for_events([t.event for t in self.events])
+
+        for t in self.events:
+            kwargs = t.kwargs
+            program = t.program
+            invoker_code, data = self.invoker_codes[program.name][tuple(kwargs)]
+            evt = t.event
+
+            types = {}
+            param_dict = {}
+
+            for key, value in kwargs.items():
+                types[key] = value.dtype
+                param_dict[key] = value
+
+            for key, value in program.arg_dict.items():
+                if key not in param_dict:
+                    param_dict[key] = None
+
+            # execute integer argument generation code from wrapper
+            exec(invoker_code, param_dict)
+
+            # get statistics
+            program = lp.add_and_infer_dtypes(program, types)
+            op_map = lp.get_op_map(program, subgroup_size='guess')
+            mem_map = lp.get_mem_access_map(program, subgroup_size='guess')
+
+            f32op_count = op_map.filter_by(dtype=[np.float32]).eval_and_sum(param_dict)
+            f64op_count = op_map.filter_by(dtype=[np.float64]).eval_and_sum(param_dict)
+
+            mem32_count = mem_map.filter_by(dtype=[np.float32]).eval_and_sum(param_dict)
+            mem64_count = mem_map.filter_by(dtype=[np.float64]).eval_and_sum(param_dict)
```

Comment:
Doing this repeatedly will be horrendously expensive without caching.

Reply:
Should be fixed in the new version.